parsers/BD.py

```diff
@@ -1,299 +1,295 @@
     try:
-        dt_aware = target_datetime.replace(
-            hour=int(hour), minute=int(minute), tzinfo="Asia/Dhaka"
-        )
+        return float(entry)
     except ValueError:
-        # 24:00 present in data, requires shifting forward.
-        dt_aware = target_datetime.shift(days=+1)
-
-    return dt_aware
-
-
-def old_format_converter(df):
-    """Returns a dataframe."""
-
-    df = df.dropna(axis="columns", thresh=20)
-    df = df.dropna(axis="index", thresh=4)
-    df["TIME"] = df["TIME"].astype(str)
-    df = df.reset_index(drop=True)
-    df = df.set_index("TIME")
-
-    return df
-
-
-def new_format_converter(df, logger: Logger):
-    """Returns a dataframe."""
-
-    df = df.rename(columns={"Plant Name": "Hour"})
-    df["Hour"] = df["Hour"].astype(str)
-    df = df.reset_index(drop=True)
-    df = df.set_index("Hour")
-    total_index = df.columns.get_loc("Total (MW)")
-    df = df.iloc[:, total_index:]
-
-    try:
-        time_index = df.index.get_loc("24:00")
-    except KeyError:
         raise ParserException(
-            "BD.py", "Structure of xlsm file for BD has altered, unable to parse.", "BD"
+            parser="BD.py",
+            message=(f'Failed to parse entry: "{entry}" to float in table.'),
         )
 
-    df = df.iloc[: time_index + 1]
 
-    # check for new columns
-    if df.shape[1] != 12:
-        logger.warning(
-            "New data columns may be present xlsm file.", extra={"key": "BD"}
-        )
-
-    return df
-
-
-def production_processer(df, target_datetime, old_format=False) -> list:
+def parse_table_body(table_body: Tag):
     """
-    Takes dataframe and extracts all production data and timestamps.
-    Returns a list of 2 element tuples in form (dict, arrow object).
+    Parse the table body returned by the URL.
+    Returns a list of rows represented by dicts.
     """
 
-    if old_format:
-        MAPPING = OLD_GENERATION_MAPPING
-    else:
-        MAPPING = GENERATION_MAPPING
-
-    processed_data = []
-    for index, row in df.iterrows():
-        production = row.to_dict()
-        dt = timestamp_converter(index, target_datetime)
-
-        mapped_production = {
-            MAPPING[k]: v for k, v in production.items() if k in MAPPING
-        }
-        mapped_production["gas"] = mapped_production.pop(
-            "gas_public"
-        ) + mapped_production.pop("gas_private")
-        mapped_production["oil"] = mapped_production.pop(
-            "oil_public"
-        ) + mapped_production.pop("oil_private")
-        processed_data.append((mapped_production, dt))
+    rows = table_body.find_all("tr")
+    row_data = []
+
+    for row in rows:
+        row_items = row.find_all("td")
+        row_items = [item.text.strip() for item in row_items]
+
+        row_dict = dict()
+
+        # date and time in [0]; [1] are DD-MM-YYYY; hh:mm:ss
+        row_dict["time"] = arrow.get(
+            row_items[0] + row_items[1], "DD-MM-YYYY" + "HH:mm:ss", tzinfo=tz
+        ).datetime
```

Comment:
This currently fails when fetching really early data becasue the time is recorded as "HH:mm" not "HH:mm:ss" this should be updated to handle both senarions. I would try and parse the date seperatly and then the time seperatly and merge those. Preferable using the native datetime object but this is not a requirement.

Reply:
Wow well spotted :) This HH:mm format seems to be present just for a few hours in the first days the data is present. 40d0757